In [1]:
import os
dircs=os.listdir('../input/combination/combine/train/')
test_dircs=os.listdir('../input/combination/combine/test/')

In [2]:
X_train_directory=[]
y_train_directory=[]
for directory in dircs:
    image_paths=os.listdir(f'../input/combination/combine/train/{directory}')
    for img_name in image_paths:
        X_train_directory.append(f'../input/combination/combine/train/{directory}/{img_name}')
        y_train_directory.append(directory)

In [3]:
X_test_directory=[]
y_test_directory=[]
for directory in dircs:
    image_paths=os.listdir(f'../input/combination/combine/test/{directory}')
    for img_name in image_paths:
        X_test_directory.append(f'../input/combination/combine/test/{directory}/{img_name}')
        y_test_directory.append(directory)

In [4]:
df={"trainimages":X_train_directory,"trainlabels":y_train_directory}


In [5]:
df2={"testimages":X_test_directory,"testlabels":y_test_directory}


In [6]:
import pandas as pd
import numpy as np
df=pd.DataFrame(df)
df2=pd.DataFrame(df2)

In [7]:
from sklearn.preprocessing import OneHotEncoder
df = df.sample(frac=1).reset_index(drop=True)
onehot=OneHotEncoder()
lab=onehot.fit_transform(df['trainlabels'].values.reshape(-1,1)).toarray()
images=df['trainimages']
len(lab)

1381

In [8]:
from sklearn.preprocessing import OneHotEncoder
df2 = df2.sample(frac=1).reset_index(drop=True)
onehot=OneHotEncoder()
lab_test=onehot.fit_transform(df2['testlabels'].values.reshape(-1,1)).toarray()
images_test=df2['testimages']
len(lab_test)

171

In [9]:
import cv2
import matplotlib.pyplot as plt
imgs=[]
for img in images:
    if img.startswith("../input/combination/combine/train/white spot/"):
        other_imag = cv2.imread(img)

        other_frame = cv2.resize(other_imag,(224,224))
        kernel = np.array([[0, -1, 0],
               [-1, 5,-1],
               [0, -1, 0]])
        image_sharp = cv2.filter2D(src=other_frame, ddepth=-1, kernel=kernel)
        hsv = cv2.cvtColor(image_sharp, cv2.COLOR_BGR2HSV)

        # define range of white color in HSV
        # change it according to your need !
        lower_white = np.array([0,0,168], dtype=np.uint8)
        upper_white = np.array([172,111,255],dtype=np.uint8)

        # Threshold the HSV image to get only white colors
        mask = cv2.inRange(hsv, lower_white, upper_white)
        # Bitwise-AND mask and original image
        res = cv2.bitwise_and(image_sharp,image_sharp, mask= mask)
        imgs.append(res)
         
    elif img.startswith("../input/combination/combine/train/red spot/"):
        imag = cv2.imread(img)

        frame = cv2.resize(imag,(224,224)) 

        kernel = np.array([[0, -1, 0],
                   [-1, 5,-1],
                   [0, -1, 0]])
        image_sharp = cv2.filter2D(src=frame, ddepth=-1, kernel=kernel)
        hsv = cv2.cvtColor(image_sharp,cv2.COLOR_BGR2HSV)

        lower_red = np.array([0,50,50])
        upper_red = np.array([10,255,255])


        #upper red
        lower_red2 = np.array([170,50,50])
        upper_red2 = np.array([180,255,255])

        mask = cv2.inRange(hsv, lower_red, upper_red)
        res = cv2.bitwise_and(image_sharp,image_sharp, mask= mask)


        mask2 = cv2.inRange(hsv, lower_red2, upper_red2)
        res2 = cv2.bitwise_and(image_sharp,image_sharp, mask= mask2)

        
        imgs.append(res2)
    elif img.startswith("../input/combination/combine/train/black spot/"):
        other_imag = cv2.imread(img)

        other_frame = cv2.resize(other_imag,(224,224))
        kernel = np.array([[0, -1, 0],
               [-1, 5,-1],
               [0, -1, 0]])
        image_sharp = cv2.filter2D(src=other_frame, ddepth=-1, kernel=kernel)
        imagehsv = cv2.cvtColor(image_sharp, cv2.COLOR_BGR2HSV)


        lower_black = np.array([0,0,0])
        upper_black = np.array([179,255,127])

        imagemask = cv2.inRange(imagehsv, lower_black, upper_black)
        result = cv2.bitwise_not(image_sharp,image_sharp, mask=imagemask)

        imgs.append(result)
    elif img.startswith("../input/combination/combine/train/fresh fish/"):
        other_imag = cv2.imread(img)

        other_frame = cv2.resize(other_imag,(224,224))
        kernel = np.array([[0, -1, 0],
               [-1, 5,-1],
               [0, -1, 0]])
        image_sharp = cv2.filter2D(src=other_frame, ddepth=-1, kernel=kernel)

        imgs.append(image_sharp)

In [10]:
# import matplotlib.pyplot as plt
# testimgs=[]
# for img in images_test:
#     if img.startswith("../input/combination/combine/test/white spot/"):
#         other_imag = cv2.imread(img)

#         other_frame = cv2.resize(other_imag,(224,224))
#         kernel = np.array([[0, -1, 0],
#                [-1, 5,-1],
#                [0, -1, 0]])
#         image_sharp = cv2.filter2D(src=other_frame, ddepth=-1, kernel=kernel)
#         hsv = cv2.cvtColor(image_sharp, cv2.COLOR_BGR2HSV)

#         # define range of white color in HSV
#         # change it according to your need !
#         lower_white = np.array([0,0,168], dtype=np.uint8)
#         upper_white = np.array([172,111,255],dtype=np.uint8)

#         # Threshold the HSV image to get only white colors
#         mask = cv2.inRange(hsv, lower_white, upper_white)
#         # Bitwise-AND mask and original image
#         res = cv2.bitwise_and(image_sharp,image_sharp, mask= mask)
#         testimgs.append(res)
         
#     elif img.startswith("../input/combination/combine/test/red spot/"):
#         imag = cv2.imread(img)

#         frame = cv2.resize(imag,(224,224)) 

#         kernel = np.array([[0, -1, 0],
#                    [-1, 5,-1],
#                    [0, -1, 0]])
#         image_sharp = cv2.filter2D(src=frame, ddepth=-1, kernel=kernel)
#         hsv = cv2.cvtColor(image_sharp,cv2.COLOR_BGR2HSV)

#         lower_red = np.array([0,50,50])
#         upper_red = np.array([10,255,255])


#         #upper red
#         lower_red2 = np.array([170,50,50])
#         upper_red2 = np.array([180,255,255])

#         mask = cv2.inRange(hsv, lower_red, upper_red)
#         res = cv2.bitwise_and(image_sharp,image_sharp, mask= mask)


#         mask2 = cv2.inRange(hsv, lower_red2, upper_red2)
#         res2 = cv2.bitwise_and(image_sharp,image_sharp, mask= mask2)

        

#         testimgs.append(res2)
#     elif img.startswith("../input/combination/combine/test/black spot/"):
#         other_imag = cv2.imread(img)

#         other_frame = cv2.resize(other_imag,(224,224))
#         kernel = np.array([[0, -1, 0],
#                [-1, 5,-1],
#                [0, -1, 0]])
#         image_sharp = cv2.filter2D(src=other_frame, ddepth=-1, kernel=kernel)
#         imagehsv = cv2.cvtColor(image_sharp, cv2.COLOR_BGR2HSV)


#         lower_black = np.array([0,0,0])
#         upper_black = np.array([179,255,127])

#         imagemask = cv2.inRange(imagehsv, lower_black, upper_black)
#         result = cv2.bitwise_not(image_sharp,image_sharp, mask=imagemask)

#         testimgs.append(result)
#     elif img.startswith("../input/combination/combine/test/fresh fish/"):
#         other_imag = cv2.imread(img)

#         other_frame = cv2.resize(other_imag,(224,224))
#         kernel = np.array([[0, -1, 0],
#                [-1, 5,-1],
#                [0, -1, 0]])
#         image_sharp = cv2.filter2D(src=other_frame, ddepth=-1, kernel=kernel)

#         testimgs.append(image_sharp)

In [11]:
X_train=np.array(imgs)
y_train=np.array(lab)
# X_test=np.array(testimgs)
# y_test=np.array(lab_test)

In [12]:
import numpy as np
np.random.seed(32)

In [13]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib.image as mimg
%matplotlib inline
plt.rcParams["figure.figsize"] = (10,7)
from PIL import Image
from scipy import misc

import os

from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix


# DEEP LEARNING IMPORTS
from keras.models import Sequential, load_model
from keras.layers import Dense, Conv2D, Activation, Dropout, Flatten, MaxPooling2D
from keras.callbacks import ModelCheckpoint, EarlyStopping

In [14]:
train_x, test_x, train_y, test_y = train_test_split(X_train, y_train, random_state = 32, 
                                                   test_size=0.25)

In [15]:

X_trainnorm=train_x/255.0
X_testnorm=test_x/255.0

In [16]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from keras.models import Sequential
from keras.layers import Dense, BatchNormalization, Dropout
from tensorflow.keras.optimizers import Adam, SGD, RMSprop, Adadelta, Adagrad, Adamax, Nadam, Ftrl
from keras.callbacks import EarlyStopping, ModelCheckpoint
from keras.wrappers.scikit_learn import KerasClassifier
from math import floor
from sklearn.metrics import make_scorer, accuracy_score
from bayes_opt import BayesianOptimization
from sklearn.model_selection import StratifiedKFold
from keras.layers import LeakyReLU
LeakyReLU = LeakyReLU(alpha=0.1)
import warnings
warnings.filterwarnings('ignore')
pd.set_option("display.max_columns", None)

In [17]:
score_acc = make_scorer(accuracy_score)

In [18]:
y_train_norm=np.argmax(train_y,axis=1)

In [19]:
y_train_norm.shape

(1035,)

In [20]:
from tensorflow import keras
from tensorflow.keras import layers
from kerastuner.tuners import RandomSearch

In [21]:
# def build_model(hp):
#     # create model object
#     model = keras.Sequential([
#     #adding first convolutional layer    
#     keras.layers.Conv2D(
#         #adding filter 
#         filters=hp.Int('conv_1_filter', min_value=64, max_value=256, step=32,default=128),
#         # adding filter size or kernel size
#         kernel_size=hp.Choice('conv_1_kernel', values = [3]),
#         #activation function
#         activation='relu',
#         input_shape=(224,224,3)),
#     # adding second convolutional layer 
   
#     # adding flatten layer    
#     keras.layers.Flatten(),
#     # adding dense layer    
# #     keras.layers.Dense(
# #         units=352,
# #         activation='relu'
# #     ),
#     # output layer    
#     keras.layers.Dense(4, activation='softmax')
#     ])
#     #compilation of model
#     model.compile(optimizer=keras.optimizers.Adam(0.0001),loss='categorical_crossentropy',metrics=['accuracy'])
#     return model

In [22]:
import keras_tuner as kt
class MyHyperModel(kt.HyperModel):
    def build(self, hp):
        model = keras.Sequential()
        
        model.add(layers.Flatten())
        model.add(
            layers.Dense(
                units=352,
                activation="relu",
            )
        )
        model.add(layers.Dense(4, activation="softmax"))
        model.compile(
            optimizer="adam", loss="categorical_crossentropy", metrics=["accuracy"],
        )
        return model

    def fit(self, hp, model, *args, **kwargs):
        return model.fit(
            *args,
            epochs=hp.Choice("epochs", [30, 50,100]),
            batch_size=hp.Choice("batch_size", [16]),
            **kwargs,
        )

tuner = kt.RandomSearch(
    MyHyperModel(),
    objective="val_accuracy",
    max_trials=3,
    overwrite=True,
    directory="my_dir",
    project_name="tune_hypermodel",
)

2022-11-24 20:38:57.131139: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-24 20:38:57.132155: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-24 20:38:57.439725: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-24 20:38:57.440672: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-24 20:38:57.441609: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from S

In [23]:
import tensorflow as tf
tuner.search(X_trainnorm, train_y,validation_data=(X_testnorm, test_y), callbacks=[tf.keras.callbacks.EarlyStopping('val_loss', patience=3)])

Trial 3 Complete [00h 00m 29s]
val_accuracy: 0.7167630195617676

Best val_accuracy So Far: 0.7167630195617676
Total elapsed time: 00h 01m 37s


In [27]:
tuner.results_summary()

Results summary
Results in my_dir/tune_hypermodel
Showing 10 best trials
Trial summary
Hyperparameters:
epochs: 100
batch_size: 16
Score: 0.7167630195617676
Trial summary
Hyperparameters:
epochs: 50
batch_size: 16
Score: 0.7023121118545532
Trial summary
Hyperparameters:
epochs: 30
batch_size: 16
Score: 0.6936416029930115


In [24]:
# asadfafaf
# asfsaftuner = RandomSearch(
#     build_model,
#     objective = 'val_accuracy',
#     max_trials = 5,
#     executions_per_trial = 3,
# )

In [25]:
tuner.search_space_summary()

Search space summary
Default search space size: 2
epochs (Choice)
{'default': 30, 'conditions': [], 'values': [30, 50, 100], 'ordered': True}
batch_size (Choice)
{'default': 16, 'conditions': [], 'values': [16], 'ordered': True}


In [26]:
tuner.search(X_trainnorm, train_y, , epochs=30, callbacks=[tf.keras.callbacks.EarlyStopping('val_loss', patience=3), validation_data=(X_testnorm, test_y))

SyntaxError: invalid syntax (3259485633.py, line 1)

###Range 8-32
Results summary
Results in ./untitled_project
Showing 10 best trials
<keras_tuner.engine.objective.Objective object at 0x7f42b87267d0>
Trial summary
Hyperparameters:
conv_1_filter: 24
conv_1_kernel: 3
Score: 0.9065510630607605
Trial summary
Hyperparameters:
conv_1_filter: 28
conv_1_kernel: 5
Score: 0.9036608934402466
Trial summary
Hyperparameters:
conv_1_filter: 20
conv_1_kernel: 3
Score: 0.8988439242045084
Trial summary
Hyperparameters:
conv_1_filter: 8
conv_1_kernel: 5
Score: 0.8978805541992188
Trial summary
Hyperparameters:
conv_1_filter: 16
conv_1_kernel: 3
Score: 0.892100195089976

range 32-64

Results summary
Results in ./untitled_project
Showing 10 best trials
<keras_tuner.engine.objective.Objective object at 0x7fc272cedb50>
Trial summary
Hyperparameters:
conv_1_filter: 64
conv_1_kernel: 5
Score: 0.9315992395083109
Trial summary
Hyperparameters:
conv_1_filter: 48
conv_1_kernel: 3
Score: 0.9277456601460775
Trial summary
Hyperparameters:
conv_1_filter: 32
conv_1_kernel: 5
Score: 0.9219653209050497
Trial summary
Hyperparameters:
conv_1_filter: 64
conv_1_kernel: 3
Score: 0.9190751512845358
Trial summary
Hyperparameters:
conv_1_filter: 48
conv_1_kernel: 5
Score: 0.910404622554779

range 64-256
Results summary
Results in ./untitled_project
Showing 10 best trials
<keras_tuner.engine.objective.Objective object at 0x7f0932fb6210>
Trial summary
Hyperparameters:
conv_1_filter: 192
conv_1_kernel: 5
Score: 0.9364161888758341
Trial summary
Hyperparameters:
conv_1_filter: 224
conv_1_kernel: 5
Score: 0.9335260192553202
Trial summary
Hyperparameters:
conv_1_filter: 192
conv_1_kernel: 3
Score: 0.9306358496348063
Trial summary
Hyperparameters:
conv_1_filter: 160
conv_1_kernel: 5
Score: 0.9248554706573486
Trial summary
Hyperparameters:
conv_1_filter: 160
conv_1_kernel: 3
Score: 0.9248554706573486

#### Batch size

Results summary
Results in my_dir/tune_hypermodel
Showing 10 best trials
<keras_tuner.engine.objective.Objective object at 0x7fe4d40fad90>
Trial summary
Hyperparameters:
batch_size: 16
Score: 0.7832369804382324
Trial summary
Hyperparameters:
batch_size: 16
Score: 0.7543352842330933
Trial summary
Hyperparameters:
batch_size: 32
Score: 0.5057803392410278

#### Epochs
Results summary
Results in my_dir/tune_hypermodel
Showing 10 best trials
<keras_tuner.engine.objective.Objective object at 0x7f2f879ef550>
Trial summary
Hyperparameters:
epochs: 100
batch_size: 16
Score: 0.7167630195617676
Trial summary
Hyperparameters:
epochs: 50
batch_size: 16
Score: 0.7023121118545532
Trial summary
Hyperparameters:
epochs: 30
batch_size: 16
Score: 0.6936416029930115

In [ ]:
tuner.results_summary()

In [ ]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense